### PR TITLE
LSNBLDR-952; fix lessons subnav

### PR DESCRIPTION
**Pull request by clhedrick**
various details were wrong in the subnav code:
* didn't check for group access
* didn't check prereq quite right
* slightly wrong way of getting top level pages

### CODE REVIEW

File: lessonbuilder/components/src/java/org/sakaiproject/lessonbuildertool/model/SimplePageToolDaoImpl.java [hunk 1]
```diff
@@ -1903,18 +1903,21 @@ public Object readSqlResultRecord(final ResultSet result) {
 	}
 
 	public List<SimplePage> getTopLevelPages(final String siteId) {
-		DetachedCriteria d = DetachedCriteria.forClass(SimplePage.class).add(Restrictions.eq("siteId", siteId))
-			.add(Restrictions.disjunction()
-				.add(Restrictions.isNull("owner"))
-				.add(Restrictions.eq("owned", true)))
-			.add(Restrictions.isNull("parent"));
+	    // set of all top level pages, actually the items pointing to them                                                                       
+	    List<SimplePageItem> sitePages =  findItemsInSite(toolManager.getCurrentPlacement().getContext());
```
Comment (ottenhoff) [inline lessonbuilder/components/src/java/org/sakaiproject/lessonbuildertool/model/SimplePageToolDaoImpl.java]: Why are you using toolManager.getCurrentPlacement().getContext() when siteId is passed to this function?

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/entityproviders/LessonsEntityProvider.java [hunk 2]
```diff
@@ -434,16 +434,27 @@ public String getPrerequisiteDataForSubNav(final EntityView view, final Map<Stri
 
 		for (SimplePage topLevelPage : topLevelPages) {
 			final List<SimplePageItem> itemsOnPage = simplePageToolDao.findItemsOnPage(topLevelPage.getPageId());
+			final JSONObject pageData = new JSONObject();
 			final JSONArray inaccessibleItems = new JSONArray();
+			final JSONArray invisibleItems = new JSONArray();
+
+			if (!itemsOnPage.isEmpty())
+				simplePageBean = makeSimplePageBean(null, siteId, itemsOnPage.get(0));
```
Comment (ottenhoff) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/entityproviders/LessonsEntityProvider.java]: We need to use brackets in Sakai when we don't use consistent tabs and spaces

**Comment by clhedrick** [comment]
You’re right. I was copying code. I’ll fix it.

On Mar 1, 2018, at 5:00:36 PM, Sam Ottenhoff <notifications@github.com<mailto:notifications@github.com>> wrote:


@ottenhoff commented on this pull request.

________________________________

In lessonbuilder/components/src/java/org/sakaiproject/lessonbuildertool/model/SimplePageToolDaoImpl.java<https://github.com/sakaiproject/sakai/pull/5362#discussion_r171707680>:

> @@ -1903,18 +1903,21 @@ public Object readSqlResultRecord(final ResultSet result) {
        }

        public List<SimplePage> getTopLevelPages(final String siteId) {
-               DetachedCriteria d = DetachedCriteria.forClass(SimplePage.class).add(Restrictions.eq("siteId", siteId))
-                       .add(Restrictions.disjunction()
-                               .add(Restrictions.isNull("owner"))
-                               .add(Restrictions.eq("owned", true)))
-                       .add(Restrictions.isNull("parent"));
+           // set of all top level pages, actually the items pointing to them
+           List<SimplePageItem> sitePages =  findItemsInSite(toolManager.getCurrentPlacement().getContext());


Why are you using toolManager.getCurrentPlacement().getContext() when siteId is passed to this function?

File: lessonbuilder/components/src/java/org/sakaiproject/lessonbuildertool/model/SimplePageToolDaoImpl.java (hunk 1 above)
Comment (payten) [inline lessonbuilder/components/src/java/org/sakaiproject/lessonbuildertool/model/SimplePageToolDaoImpl.java]: Thanks for fixing things up @clhedrick.

My only query is regarding the use of `findItemsInSite`. Performance is important for this feature, as we try to make the subnav initialisation relatively seamless upon page load.  Does the change to use this increase the performance overhead enough for us to worry?

Comment (ottenhoff) [reply lessonbuilder/components/src/java/org/sakaiproject/lessonbuildertool/model/SimplePageToolDaoImpl.java]: We should worry about this change! This is going to change this lookup from one query into possibly hundreds or thousands of queries (site => pages => items), right?!

**Comment by clhedrick** [comment]
I’ve already rewritten it. I’m about to check it in. No, it’s not so bad, but it’s unnecessarily complex.

finditemsinsite (which is actually find all top-level items in site) does

* find all site pages. I assume one query
* find all lessons pages with toolid properties pointing to one of them. One query.
* For each page find the item pointing to it. done with a single specialized query

What I did unnecessarily was call that code and then look up each item to find the page. That’s unnecessary lookups, but only the number of top level pages.

In fact that’s silly. I’ve copied the code of finditemsinsite and left off the last step.

On Mar 2, 2018, at 9:49 AM, Sam Ottenhoff <notifications@github.com<mailto:notifications@github.com>> wrote:


@ottenhoff commented on this pull request.

________________________________

In lessonbuilder/components/src/java/org/sakaiproject/lessonbuildertool/model/SimplePageToolDaoImpl.java<https://github.com/sakaiproject/sakai/pull/5362#discussion_r171865989>:

> @@ -1903,18 +1903,21 @@ public Object readSqlResultRecord(final ResultSet result) {
        }

        public List<SimplePage> getTopLevelPages(final String siteId) {
-               DetachedCriteria d = DetachedCriteria.forClass(SimplePage.class).add(Restrictions.eq("siteId", siteId))
-                       .add(Restrictions.disjunction()
-                               .add(Restrictions.isNull("owner"))
-                               .add(Restrictions.eq("owned", true)))
-                       .add(Restrictions.isNull("parent"));
+           // set of all top level pages, actually the items pointing to them
+           List<SimplePageItem> sitePages =  findItemsInSite(toolManager.getCurrentPlacement().getContext());


We should worry about this change! This is going to change this lookup from one query into possibly hundreds or thousands of queries (site => pages => items), right?!

**Comment by ottenhoff** [comment]
We have looked into performance issues in findItemsInSite before. That history is here: 

  https://jira.sakaiproject.org/browse/LSNBLDR-877

Please leave future PRs open for community review. We usually leave our PRs open for 3-5 days to try to get comments from community.

**Comment by clhedrick** [comment]
Ugh. We have a rogue advisor. IsInsstructor is always set, because everyone shows as having site.upd.

**Comment by ottenhoff** [comment]
A related JIRA of advisor cleanup: https://jira.sakaiproject.org/browse/LSNBLDR-869

**Pull request merged**